src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,259 @@
+                Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Raw, ProtocolType.Icmp);
+                s.Dispose();
+                return true;
+            }
+            catch
+            {
+                return false;
+            }
+        }
+
         private void InternalSendAsync(IPAddress address, byte[] buffer, int timeout, PingOptions options)
         {
             AsyncOperation asyncOp = _asyncOp;
             SendOrPostCallback callback = _onPingCompletedDelegate;
 
-            // TODO: Implement this (#2487)
+            PingReply pr = null;
+            Exception pingException = null;
+
+            try
+            {
+                if (s_canUseRawSockets)
+                {
+                    pr = SendIcmpPingOverRawSocket(address, buffer, timeout, options);
+                }
+                else
+                {
+                    pr = SendWithPingUtility(address, buffer, timeout, options);
+                }
+            }
+            catch (Exception e)
+            {
+                pingException = e;
+            }
 
-            PingException pe = new PingException(SR.net_ping, new PlatformNotSupportedException());
+            // At this point, either PR has a real PingReply in it, or pingException has an Exception in it.
             var ea = new PingCompletedEventArgs(
-                new PingReply(address, default(PingOptions), default(IPStatus), default(long), buffer), 
-                pe, 
-                false, 
+                pr,
+                pingException,
+                false,
                 asyncOp.UserSuppliedState);
 
             Finish();
             asyncOp.PostOperationCompleted(callback, ea);
         }
 
+        private unsafe PingReply SendIcmpPingOverRawSocket(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        {
+            EndPoint endPoint = new IPEndPoint(address, 0);
+
+            bool isIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            ProtocolType protocolType = isIpv4 ? ProtocolType.Icmp : ProtocolType.IcmpV6;
+            // Use the current thread's ID as the identifier.
+            ushort identifier = (ushort)Environment.CurrentManagedThreadId;
+            IcmpHeader header = new IcmpHeader()
+            {
+                Type = isIpv4 ? (byte)Icmpv4MessageConstants.EchoRequest : (byte)Icmpv6MessageConstants.EchoRequest,
+                Code = 0,
+                HeaderChecksum = 0,
+                Identifier = identifier,
+                SequenceNumber = 0,
+            };
+            byte[] sendBuffer = CreateSendMessageBuffer(header, buffer);
+
+            using (Socket socket = new Socket(address.AddressFamily, SocketType.Raw, protocolType))
+            {
+                socket.ReceiveTimeout = timeout;
+                socket.SendTimeout = timeout;
+                // Setting Socket.DontFragment and .Ttl is not supported on Unix, so ignore the PingOptions parameter.
```

Comment:
That's one possibility. @pgavlin: Is this a platform restriction, or could we eventually support setting those values? Right now they both throw PlatformNotSupportedException's.